### PR TITLE
Inspection middleware with jsonpickle

### LINKED ISSUE #334: Port Inspection Middleware
Parity with C# and JS

### PR DESCRIPTION
**Pull request by axelsrz**
Inspection Middleware leveraging jsonpickle.

fixes #334

### CODE REVIEW

File: libraries/botbuilder-core/botbuilder/core/bot_state.py [hunk 1]
```diff
@@ -186,17 +187,23 @@ def name(self) -> str:
...
         except:
             # ask for default value from factory
-            if not default_value_factory:
+            if not default_value_or_factory:
                 return None
-            result = default_value_factory()
+            result = (
+                default_value_or_factory()
+                if callable(default_value_or_factory)
+                else deepcopy(default_value_or_factory)
```
Comment (daveta) [inline libraries/botbuilder-core/botbuilder/core/bot_state.py]: Nice

**Comment by daveta** [comment]
>     packages=["botbuilder.core", "botbuilder.core.adapters"],

inspection.

---
Refers to: libraries/botbuilder-core/setup.py:34 in 927a736. [](commit_id = 927a736afb2e1442bbedac58c42b6982dd6e46f5, deletion_comment = False)

**Comment by axelsrz** [comment]
> > ```
> > packages=["botbuilder.core", "botbuilder.core.adapters"],
> > ```
> 
> inspection.
> 
> Refers to: libraries/botbuilder-core/setup.py:34 in [927a736](https://github.com/microsoft/botbuilder-python/commit/927a736afb2e1442bbedac58c42b6982dd6e46f5). [](commit_id = [927a736](https://github.com/microsoft/botbuilder-python/commit/927a736afb2e1442bbedac58c42b6982dd6e46f5), deletion_comment = False)

Done

File: libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    async def _invoke_send(
+        self, context: TurnContext, session: InspectionSession, activity: Activity
+    ) -> Any:
+        if await session.send(activity):
+            return True
+
+        await self._clean_up_session(context)
+        return False
+
+    async def _clean_up_session(self, context: TurnContext) -> Any:
```
Comment (daveta) [inline libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py]: >Any [](start = 63, length = 3)

Should this be -> None?

File: libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            context.activity.conversation.id
+        )
+        if conversation_reference:
+            return InspectionSession(conversation_reference, self.credentials)
+
+        return None
+
+    async def _invoke_send(
+        self, context: TurnContext, session: InspectionSession, activity: Activity
+    ) -> Any:
```
Comment (daveta) [inline libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py]: >Any [](start = 9, length = 3)

Should this be bool?

File: libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        await context.send_activity(
+            make_command_activity(
+                f"{InspectionMiddleware._COMMAND} attach {session_id}"
+            )
+        )
+        await self.inspection_state.save_changes(context, False)
+
+    async def process_attach_command(
+        self, context: TurnContext, session_id: str
+    ) -> Any:
```
Comment (daveta) [inline libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py]: >Any [](start = 9, length = 3)

None?

File: libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py (hunk 3 above)
Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py]: Corrected

File: libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py (hunk 2 above)
Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py]: Corrected

File: libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py (hunk 4 above)
Comment (axelsrz) [reply libraries/botbuilder-core/botbuilder/core/inspection/inspection_middleware.py]: Solved

**Pull request merged**